File: sdks/typescript/packages/observability/src/collector.ts
import type { PaymentEvent, PaymentEventFilter, PaymentMetrics } from "./types";

/**
 * Default maximum number of events retained in the ring buffer.
 */
const DEFAULT_MAX_SIZE = 10_000;

/**
 * Collects payment events and computes aggregate metrics.
 *
 * Uses an in-memory ring buffer with configurable maximum size.
 * When the buffer is full, the oldest events are discarded.
 */
export class PaymentEventCollector {
  private buffer: PaymentEvent[];
  private head: number;
  private count: number;
  private readonly maxSize: number;

  /**
   * Create a new PaymentEventCollector.
   *
   * @param maxSize - Maximum number of events to retain (default: 10000)
   */
  constructor(maxSize: number = DEFAULT_MAX_SIZE) {
    if (maxSize < 1) {
      throw new Error("maxSize must be at least 1");
    }
    this.maxSize = maxSize;
    this.buffer = new Array(maxSize);
    this.head = 0;
    this.count = 0;
  }

  /**
   * Record a payment event.
   *
   * @param event - The payment event to record
   */
  record(event: PaymentEvent): void {
    this.buffer[this.head] = event;
    this.head = (this.head + 1) % this.maxSize;
    if (this.count < this.maxSize) {
      this.count++;
    }
  }

  /**
   * Get all recorded events in chronological order, optionally filtered.
   *
   * @param filter - Optional filter criteria
   * @returns Array of matching events
   */
  getEvents(filter?: PaymentEventFilter): PaymentEvent[] {
    let events = this.getAllEventsOrdered();

    if (!filter) {
      return events;
    }

    if (filter.type) {
      events = events.filter((e) => e.type === filter.type);
    }
    if (filter.paymentId) {
      events = events.filter((e) => e.paymentId === filter.paymentId);
    }
    if (filter.network) {
      events = events.filter((e) => e.network === filter.network);
    }
    if (filter.after !== undefined) {
      events = events.filter((e) => e.timestamp > filter.after!);
    }
    if (filter.before !== undefined) {
      events = events.filter((e) => e.timestamp < filter.before!);
    }
    if (filter.limit !== undefined && filter.limit > 0) {
      events = events.slice(0, filter.limit);
    }

    return events;
  }

  /**
   * Compute aggregate metrics from all collected events.
   *
   * @returns Computed payment metrics
   */
  getMetrics(): PaymentMetrics {
    const events = this.getAllEventsOrdered();

    const metrics: PaymentMetrics = {
      totalAttempted: 0,
      totalSuccessful: 0,
      totalFailed: 0,
      avgVerifyLatencyMs: 0,
      avgSettleLatencyMs: 0,
      amountByNetwork: {},
      countByNetwork: {},
      failureReasons: {},
    };

    // Group events by paymentId for latency calculations
    const flowEvents = new Map<string, PaymentEvent[]>();
    for (const event of events) {
      if (!flowEvents.has(event.paymentId)) {
        flowEvents.set(event.paymentId, []);
      }
      flowEvents.get(event.paymentId)!.push(event);
    }

    let verifyLatencySum = 0;
    let verifyLatencyCount = 0;
    let settleLatencySum = 0;
    let settleLatencyCount = 0;

    for (const [, flowEvts] of flowEvents) {
      const hasRequested = flowEvts.some((e) => e.type === "payment.requested");
      const hasCompleted = flowEvts.some((e) => e.type === "payment.completed");
      const hasFailed = flowEvts.some((e) => e.type === "payment.failed");

      if (hasRequested) {
        metrics.totalAttempted++;
      }
      if (hasCompleted) {
        metrics.totalSuccessful++;
      }
      if (hasFailed) {
        metrics.totalFailed++;
        // Collect failure reasons
        for (const evt of flowEvts) {
          if (evt.type === "payment.failed" && evt.error) {
            metrics.failureReasons[evt.error] = (metrics.failureReasons[evt.error] ?? 0) + 1;
          }
        }
      }

      // Compute verify latency: submitted -> verified
      const submitted = flowEvts.find((e) => e.type === "payment.submitted");
      const verified = flowEvts.find((e) => e.type === "payment.verified");
      if (submitted && verified) {
        verifyLatencySum += verified.timestamp - submitted.timestamp;
        verifyLatencyCount++;
      }

      // Compute settle latency: verified -> settled
      const settled = flowEvts.find((e) => e.type === "payment.settled");
      if (verified && settled) {
        settleLatencySum += settled.timestamp - verified.timestamp;
        settleLatencyCount++;
      }

      // Track amounts and counts by network
      const completedEvt = flowEvts.find((e) => e.type === "payment.completed");
      if (completedEvt?.network) {
        const network = completedEvt.network;
        metrics.countByNetwork[network] = (metrics.countByNetwork[network] ?? 0) + 1;
        if (completedEvt.amount) {
          const currentAmount = metrics.amountByNetwork[network] ?? 0n;
          metrics.amountByNetwork[network] = currentAmount + BigInt(completedEvt.amount);
        }
      }
    }

    metrics.avgVerifyLatencyMs = verifyLatencyCount > 0 ? verifyLatencySum / verifyLatencyCount : 0;
    metrics.avgSettleLatencyMs = settleLatencyCount > 0 ? settleLatencySum / settleLatencyCount : 0;

    return metrics;
  }

  /**
   * Get the number of events currently stored.
   */
  get size(): number {
    return this.count;
  }

  /**
   * Clear all recorded events.
   */
  clear(): void {
    this.buffer = new Array(this.maxSize);
    this.head = 0;
    this.count = 0;
  }

  /**
   * Get all events in chronological order from the ring buffer.
   */
  private getAllEventsOrdered(): PaymentEvent[] {
    if (this.count === 0) {
      return [];
    }

    const result: PaymentEvent[] = [];
    const start = this.count < this.maxSize ? 0 : this.head;
    for (let i = 0; i < this.count; i++) {
      const idx = (start + i) % this.maxSize;
      result.push(this.buffer[idx]);
    }
    return result;
  }
}

File: sdks/go/observability/collector.go
package observability

import (
	"math/big"
	"sync"
	"time"
)

// DefaultBufferSize is the default ring buffer capacity.
const DefaultBufferSize = 1000

// Collector records payment events in a thread-safe ring buffer and computes aggregate metrics.
type Collector struct {
	mu       sync.RWMutex
	events   []PaymentEvent
	capacity int
	head     int // next write position
	count    int // total events stored (capped at capacity)
}

// NewCollector creates a Collector with the given ring buffer capacity.
// If capacity <= 0, DefaultBufferSize is used.
func NewCollector(capacity int) *Collector {
	if capacity <= 0 {
		capacity = DefaultBufferSize
	}
	return &Collector{
		events:   make([]PaymentEvent, capacity),
		capacity: capacity,
	}
}

// Record adds a payment event to the ring buffer.
// If the buffer is full the oldest event is overwritten.
func (c *Collector) Record(event PaymentEvent) {
	if event.Timestamp.IsZero() {
		event.Timestamp = time.Now()
	}
	c.mu.Lock()
	c.events[c.head] = event
	c.head = (c.head + 1) % c.capacity
	if c.count < c.capacity {
		c.count++
	}
	c.mu.Unlock()
}

// GetEvents returns events matching the filter. A zero-value filter matches everything.
func (c *Collector) GetEvents(filter EventFilter) []PaymentEvent {
	c.mu.RLock()
	defer c.mu.RUnlock()

	var result []PaymentEvent
	for _, ev := range c.snapshot() {
		if filter.PaymentID != "" && ev.PaymentID != filter.PaymentID {
			continue
		}
		if filter.Type != "" && ev.Type != filter.Type {
			continue
		}
		if filter.Network != "" && ev.Network != filter.Network {
			continue
		}
		if !filter.Since.IsZero() && ev.Timestamp.Before(filter.Since) {
			continue
		}
		result = append(result, ev)
		if filter.Limit > 0 && len(result) >= filter.Limit {
			break
		}
	}
	return result
}

// GetMetrics computes aggregate metrics from all events in the buffer.
func (c *Collector) GetMetrics() PaymentMetrics {
	c.mu.RLock()
	defer c.mu.RUnlock()

	m := PaymentMetrics{
		AmountByNetwork: make(map[string]*big.Int),
		CountByNetwork:  make(map[string]int),
		FailureReasons:  make(map[string]int),
	}

	// Track seen payment IDs per type for dedup.
	attempted := make(map[string]bool)
	successful := make(map[string]bool)
	failed := make(map[string]bool)

	var verifyLatencies, settleLatencies []int64

	for _, ev := range c.snapshot() {
		switch ev.Type {
		case EventRequested:
			attempted[ev.PaymentID] = true
		case EventCompleted:
			successful[ev.PaymentID] = true
		case EventFailed:
			failed[ev.PaymentID] = true
			if ev.Error != "" {
				m.FailureReasons[ev.Error]++
			}
		case EventVerified:
			if ev.DurationMs > 0 {
				verifyLatencies = append(verifyLatencies, ev.DurationMs)
			}
		case EventSettled:
			if ev.DurationMs > 0 {
				settleLatencies = append(settleLatencies, ev.DurationMs)
			}
		}

		// Accumulate amounts by network for completed events.
		if ev.Type == EventCompleted && ev.Network != "" && ev.Amount != "" {
			amt, ok := new(big.Int).SetString(ev.Amount, 10)
			if ok {
				if m.AmountByNetwork[ev.Network] == nil {
					m.AmountByNetwork[ev.Network] = new(big.Int)
				}
				m.AmountByNetwork[ev.Network].Add(m.AmountByNetwork[ev.Network], amt)
			}
			m.CountByNetwork[ev.Network]++
		}
	}

	m.TotalAttempted = len(attempted)
	m.TotalSuccessful = len(successful)
	m.TotalFailed = len(failed)
	m.AvgVerifyLatencyMs = avg(verifyLatencies)
	m.AvgSettleLatencyMs = avg(settleLatencies)

	return m
}

// Len returns the number of events currently in the buffer.
func (c *Collector) Len() int {
	c.mu.RLock()
	defer c.mu.RUnlock()
	return c.count
}

// Clear removes all events from the buffer.
func (c *Collector) Clear() {
	c.mu.Lock()
	c.events = make([]PaymentEvent, c.capacity)
	c.head = 0
	c.count = 0
	c.mu.Unlock()
}

// snapshot returns events in chronological order (oldest first). Caller must hold mu.
func (c *Collector) snapshot() []PaymentEvent {
	if c.count == 0 {
		return nil
	}
	result := make([]PaymentEvent, 0, c.count)
	start := 0
	if c.count == c.capacity {
		start = c.head // oldest element when buffer is full
	}
	for i := 0; i < c.count; i++ {
		idx := (start + i) % c.capacity
		result = append(result, c.events[idx])
	}
	return result
}

func avg(vals []int64) float64 {
	if len(vals) == 0 {
		return 0
	}
	var sum int64
	for _, v := range vals {
		sum += v
	}
	return float64(sum) / float64(len(vals))
}
